dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/GeneImporter.java

```diff
@@ -57,32 +63,50 @@ public ImportSource getSource() {
     return ImportSource.GENES;
   }
 
+  /**
+   * Main pipeline execution for creating gene model. Calls all readers to pre-compute hashmaps of relevant information
+   * before streaming GTF file for construction of gene model skeleton.
+   */
+  @SneakyThrows
   @Override
   public void execute() {
-    execute(all());
-  }
-
-  @SneakyThrows
-  public void execute(GeneFilter geneFilter) {
     val watch = createStarted();
 
-    log.info("Reading genes BSON stream from {}...", genesBsonUri);
-    val genesBson = readGenesBson();
+    // TODO: Investigate ordering to see if gene.txt does not need to be read twice.
+    log.info("Doing Ensembl Data Joining...");
+    val idMap = IdReader.getIdMap();
+    val nameMap = NameReader.readXrefDisplay();
+    val synMap = SynonymReader.getSynonymMap(idMap);
+    val canonicalMap = GeneReader.canonicalMap();
+    val transMap = TransReader.joinTrans();
+    val pfeatures = DomainReader.createProteinFeatures(transMap);
+    val externalIds = ExternalReader.externalIds();
+    log.info("... Done Ensemble Data Joining!");
+
+    log.info("Starting ASN.1 Import from NCBI.");
+    val asnReader = new ASNReader();
+    val summaryMap = asnReader.callGene2Xml();
+    log.info("Staged {} Summaries from NCBI.", summaryMap.size());
+
+    log.info("Reading genes gzip stream from {}...", gtfUri);
+    val geneReader = getReader();
 
     log.info("Writing genes to {}...", mongoUri);
-    writeGenes(genesBson, geneFilter);
+    val writer =
+        new GeneWriter(mongoUri, geneReader, summaryMap, nameMap, synMap, canonicalMap, pfeatures, externalIds);
+    writer.consumeGenes();
```

Comment:
I find it a bit strange that a writer be doing the control flow. To me there should the following actors:

- Importer: Encapsulates the whole import sequence
- Readers: Individual readers for each data source / reference type
- Writer: Writer for the final output stream
- Model(s): Results of the readers and / or the output of the writer
- Processer: Controls the reading and writing process on a per record basis.